### PR TITLE
Change a Work's default title on admin side

### PR DESCRIPTION
**Pull request by mrsalixor**
Takes care of issue #370 by letting an admin chose which title is default on a Work !

### CODE REVIEW

File: mangaki/mangaki/utils/anidb.py [hunk 1]
```diff
@@ -186,23 +188,21 @@ def get_or_update_work(self,
 
         # Handling of titles
         main_title = None
-        synonyms = {}
-        titles = {}
+        titles = []
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: Nice catch :D.

File: mangaki/mangaki/utils/anidb.py [hunk 2]
```diff
@@ -93,7 +93,7 @@ def lang_map(self) -> Dict[str, ExtLanguage]:
 
     @cached_property
     def unknown_language(self) -> ExtLanguage:
-        return ExtLanguage.objects.get(source='anidb', ext_lang='x-unk')
+        return ExtLanguage.objects.filter(source='anidb', ext_lang='x-unk').first()
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: What happened on your side to make this change?
Theorically, the couple `('anidb', 'x-unk')` should be unique in `ExtLanguage` :/.

Comment (mrsalixor) [inline mangaki/mangaki/utils/anidb.py]: The issue was that if there's no result, then an error is raised : `mangaki.models.DoesNotExist: ExtLanguage matching query does not exist`. It also seems that the `ExtLanguage` database does not fill up (it's empty on my local server), which might be why using a `get` fails.

Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: Theorically a data migration should populate this.
Hmm. I'll look on my side and get back on this. Keep it as-is for now so you can work on the PR, I'll ping for the appropriate resolution.

**Comment by mrsalixor** [comment]
It's now possible to change the default title of one or multiple works ! 🎉 
Only works with alternative titles can be modified, and duplicate titles are not shown on the form.

In terms of queries when updating titles, we have :
- 1 query in every case to retrieve the list of `(new_title, current_title, work_id)`
- 1 query per work which title changed to update the title

File: mangaki/mangaki/admin.py [hunk 3]
```diff
@@ -355,6 +358,65 @@ def refresh_work(self, request, queryset):
...
+
+            titles = WorkTitle.objects.filter(
+                pk__in=new_titles.values(), work__id__in=new_titles.keys()
+            ).values_list('title', 'work__title', 'work__id')
+
+            for new_title, current_title, work_id in titles:
+                if new_title != current_title:
+                    Work.objects.filter(pk=work_id).update(title=new_title)
+
+            self.message_user(request, 'Les titres ont bien été changés pour les œuvres sélectionnées.')
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: I'm not sure, but I think it's « changé » rather than « changés » due to the presence of « avoir » as a verb.

File: mangaki/mangaki/admin.py [hunk 4]
```diff
@@ -355,6 +358,65 @@ def refresh_work(self, request, queryset):
...
+            return None
+
+        work_titles = WorkTitle.objects.filter(work__in=queryset.values_list('pk', flat=True))
+        titles = {}
+
+        for work in queryset:
+            work_titles_for_work = work_titles.filter(work=work)
+            if len(work_titles_for_work) <= 0:
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Isn't better to load the `values_list` and do a `len` rather than before?
This way, you do one query less.

`len` will force-evaluate the queryset `work_titles_for_work` and do a `len` on the list (in most of the cases, you should use `queryset.count() == 0` and on this particular case, `queryset.exists()`, `<=` may be strange, as it's a positive integer, it should never ever be negative, this is an extra safe assumption.

File: mangaki/mangaki/admin.py [hunk 5]
```diff
@@ -355,6 +358,65 @@ def refresh_work(self, request, queryset):
...
+
+            titles[work.id] = {
+                info[0]: {
+                    'title': info[1],
+                    'language': info[2],
+                    'type': info[3] if info[1] != work.title else 'current'
+                } for info in infos
+            }
+
+        if len(titles) > 0:
```
Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Another way to write it is: `if titles:` (i.e. if the `titles` array is not empty).

File: mangaki/mangaki/admin.py (hunk 3 above)
Comment (SubaruSumeragi) [inline mangaki/mangaki/admin.py]: La conjugaison est correcte ici (l'accord vient du verbe être qui précède ton participe passé ;) )

Comment (RaitoBezarius) [inline mangaki/mangaki/admin.py]: Thanks @SubaruSumeragi ; I'm becoming more and more rusty at french :D…

File: mangaki/mangaki/tests/test_change_default_title.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+from django.test import TestCase
+from django.core.urlresolvers import reverse
+from django.contrib.auth import get_user_model
+from django.contrib import admin
+from mangaki.models import Work, Category, WorkTitle, Editor, Studio
+
+
+class MergeTest(TestCase):
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_change_default_title.py]: You meant: ChangeDefaultTitleTest ?

File: mangaki/mangaki/tests/test_change_default_title.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            admin.ACTION_CHECKBOX_NAME: self.work_ids,
+            'confirm': 1,
+            'work_ids': self.work_ids,
+            'title_ids': self.worktitle_ids
+        }
+
+        self.client.post(change_title_url, context)
+
+        self.assertTrue(Work.objects.filter(title='3-gatsu no Lion').count() > 0)
+        self.assertTrue(Work.objects.filter(title='Sound! Euphonium').count() > 0)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_change_default_title.py]: exists instead of count() > 0 ?

File: mangaki/mangaki/utils/anidb.py [hunk 8]
```diff
@@ -87,9 +87,7 @@ def lang_map(self) -> Dict[str, ExtLanguage]:
             .select_related('lang')
         )
 
-        return {
-            ext.lang.code: ext for ext in ext_langs
-        }
+        return {ext.lang.code: ext for ext in ext_langs}
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: Can you revert this change?

Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: +please

**Comment by RaitoBezarius** [comment]
Can you rebase the branch over master?
